### ISSUE TITLE
builder 0.14.10 regression (import out of order)

### ISSUE THREAD

**Issue by typhonrt** [systemjs/builder#386]
Greets Guy,

So, I believe I found another significant regression in SystemJS Builder `0.14.10`. In this case it's due to import order and how things are loaded in a static bundle. I see that a lot of changes occurred in trace.js and compile.js between `0.14.9` & `0.14.10`. Also a conceivable dependency `algorithms` used in trace.js and any replacement perhaps could be a potential source. Of course I don't have full comprehension of the deep internals of SystemJS Builder, but the algorithms "depthFirst" usage in 0.14.9 stuck out as a significant change when removed for 0.14.10 (if it's related to this area).

In #384 I mentioned a scenario with my larger web app framework and demo that it seemed like initialization code was not being run. Indeed the problem seems to be loading out of order for static bundles and not related to #384. Everything works with running in the browser via Babel, so I do think a change to how import order is handled for static bundles with Builder is the culprit. 

For my testing in the main entry point (`main.js`) I eliminated everything except two imports. In this case it's 'parseinit' followed by 'typhonUser', so the only thing in main.js file is this:

```
import 'parseinit';
import 'typhonUser'
```

`parseinit` must run first and the only shared import dependency shared between them is Parse. 

`parseinit` is super simple and just invokes the Parse initialize function. 
```
import Parse      from 'parse';
import appConfig  from 'appconfig';

Parse.initialize(appConfig.parseAppID, appConfig.parseAPIKey);
```

The imports for `typhonUser` are minimal:
```
import _          from 'underscore';
import Parse      from 'parse';
import eventbus   from 'mainEventbus';
```

In Builder 0.14.10 'typhonUser' is run first and it instantiates an instance of a class which in the constructor accesses the Parse API which is not initialized yet. 

Here is what Builder 0.14.9 outputs for the simple `main.js` listed above ('53' is `parseinit` and 'a2' is `typhonUser`). 
```
$__System.register('1', ['53', 'a2'], function (_export) {
  'use strict';

  return {
    setters: [function (_) {}, function (_a2) {}],
    execute: function () {}
  };
});
```

Here is what Builder 0.14.10 outputs for the same file ('64' is `typhonUser` and 'a2' is `parseinit`):
```
$__System.register('1', ['64', 'a2'], function (_export) {
  'use strict';

  return {
    setters: [function (_) {}, function (_a2) {}],
    execute: function () {}
  };
});
```

This seems to be a legitimate regression with Builder `0.14.10`.

**Comment by lookfirst** [comment]
On Tue, Nov 3, 2015 at 6:08 AM, Michael Leahy <notifications@github.com>
wrote:

> For my testing in my main entry point (main.js) I eliminated everything
> except two imports. In this case it's 'parseinit' followed by 'typhonUser',
> so the only thing in main.js file is this:
>
> import 'parseinit';
> import 'typhonUser'
>
> parseinit must run first and there are no dependencies between them or
> other code.
>
>
If parseinit must run before typhonUser, then that means that typhonUser
has a dependency on parseinit. You shouldn't rely on import order to
determine a dependency chain.

**Comment by typhonrt** [comment]
It doesn't have a dependency on 'parseinit'

The code I listed is from a project that worked across SystemJS Builder ~0.14.7 or 8 and broke with 0.14.10

**Comment by lookfirst** [comment]
Maybe you don't understand how dependencies work or I don't understand your problem. =) 

That said, I never rely on the order of imports and neither should you. =)

**Comment by typhonrt** [comment]
I agree it's not ideal to depend on import order due to ES6 hoisted imports. I'm dealing with a third party library that requires an initialization function to be called. It's the only thing in my framework that requires explicit initialization.

I did list out the import / dependencies in both module / files above.  Things also work in the browser with no problems. What is unclear such that you don't understand the problem?

**Comment by lookfirst** [comment]
If typhonUser depends on parseinit to execute code first (hopefully that isn't async either), then typhonUser has a "dependency" on parseinit and should declare that dependency at the top of typhonUser with an import statement pointing at parseinit.

**Comment by typhonrt** [comment]
It's not async, so that much is good; just sets API keys for a 3rd party library. 

What you are suggesting isn't generally desirable as typhonUser is coming from a separate repo that doesn't depend on the repo containing parseinit. It's hypothetically speaking a messy solution when the problem could actually be a regression in SystemJS Builder. If things were broken in the browser via Babel I'd agree that this is the solution. 

I suppose I can create an entirely separate repo that only contains 'parseinit' and also make it a main entry point for JSPM. The reason this is not great is that other 3rd parties will be building on top of my framework. To always expect 3rd parties to properly setup 'parseinit' as an explicit dependency is generally not desirable. This issue shouldn't leak beyond the bootstrapping of my own framework. 

I'm still leaning on the notion that the trace / graph aspects of Builder has changed and it's not presently in sync with Babel or the ES6 spec necessarily though this is an assumption.

**Comment by lookfirst** [comment]
Think of import statements as async loaded `<script>` tags. You can't rely on the order of execution. SystemJS builder might have been working for you in the past, but you shouldn't rely on it for execution order. Order of import statements is also a sure sign of brittle code anyway. Someone will come along and change the order and break things.

The other way around this is to export a function from typhonUser that gets initialized with parseinit.

```
import TyphonUser from 'typhonUser'
import parseInit from 'parseInit';

let user = new TyphonUser(parseInit);
```

The way you're coding things is like using a static singleton in Java to initialize things. I'd go with more of a dependency injection (guice) route where you are effectively using constructors. Your other idea of having a separate dependency for parseInit would also work.

**Comment by typhonrt** [comment]
I'm exploring the separate repo direction just for 'parseinit'. It's not the end of the world if my framework repos which depend on Parse also depend on the init repo. This should practically guarantee that 'parseinit' should be invoked before things escape the bootstrapping process where 3rd parties using my framework are concerned.

I suppose it's an open question / comment on SystemJS Builder. Should it match Babel?

**Comment by guybedford** [comment]
This output of dependencies is actually generated entirely by Babel. Are you using the same version of Babel between test cases? I'm wondering if this could be a Babel regression.

**Comment by Charlynux** [comment]
Hello,

I found a similar problem with systemjs-builder 0.14.10 on my project. 
CSS files are in the wrong order (exact revert of expected).

I create a repository to show the problem :
https://github.com/Charlynux/systemjs-order

Hope this could help.

**Comment by guybedford** [comment]
I've reverted back to post ordering in https://github.com/systemjs/builder/commit/34033865f83eb49ede7ae5c672903ee9a1de37cf, let me know if that helps.

**Comment by guybedford** [comment]
Released in 0.14.12.
